scripts: runners: bossac: Fix flash offset

nandojve opened the pull request:
* Extract code from do_run to keep clean
* Set 1200 speed when board define any extended SAM-BA bootloader (Arduino, Adafruit UF2). Keep 115200 as default to SAM-BA bootloader devices.
* Fix offset parameter and SDK checks
* Add SAM-BA bootloader documentation ~(probably is not on the right place and need guidance)~
* Update all tests

The change were tested with:

SAM4E using serial port (ROM bootloader)
SAM4L using USB (flash bootloader)

Requirements: SDK 0.12.0-beta (only SDK 0.12.0 will enable full SAM-BA support)

nandojve committed:
doc: guides: debugging: Move to flash_debug dir

The debugging section was generalized for flash and debugging tools.
The links were refactored accordly and all references were updated.
To be consistent, the directory doc/guides/debbugging was renamed to
doc/guides/flash_debug.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
scripts: runners: bossac: Fix offset parameter

The current bossac runner have multiple entries for flash offset
parameter.  Remove offset parameter from command line and rework
all infraestructure to get offset from device tree.  Add proper
verification routines to validate configurations on board and
device tree entries to fix SDK compatibility checks.

Fixes #29312.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
scripts: runners: bossac: Extract set_serial_config

Move serial settings outside do_run method to keep clean code.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
scripts: runners: bossac: Change speed by protocol

Add is_extended_samba_protocol method to verify if board is using SAM-BA
extended protocol (Arduino/Adafruit UF2).  It allows enforce serial line
speed to 1200 baud automatically.  This avoid add entries on board.cmake
file.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
scripts: runners: bossac: Extract make_bossac_cmd

Move bossac command construction outside do_run method to keep
clean code.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+
+This devices user should define:
+
+1. The flash partitions required to acomodate bootloader and image; see
+   :ref:`flash_map_api` for details.
+2. Your board's :file:`.defconfig` file must have
+   :option:`CONFIG_USE_DT_CODE_PARTITION` Kconfig option as Y in order to
+   instructs proper Zephyr code relocationg.
+3. You need build and flash the SAM-BA bootloader for your device
```
```suggestion
1. Define lash partitions required to accomodate the bootloader and application image; see
   :ref:`flash_map_api` for details.
2. Have board :file:`.defconfig` file with the
   :option:`CONFIG_USE_DT_CODE_PARTITION` Kconfig option set to ``y`` to
   instruct the build system to use these partitions for code relocation.
   This option can also be set in ``prj.conf`` or any other Kconfig fragment.
3. Build and flash the SAM-BA bootloader on the device
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+Devices with compatible SAM-BA bootloader
+-----------------------------------------
+
+This devices user should define:
```
```suggestion
For these devices, the user should:
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+
+This devices user should define:
+
+1. The flash partitions required to acomodate bootloader and image; see
+   :ref:`flash_map_api` for details.
+2. Your board's :file:`.defconfig` file must have
+   :option:`CONFIG_BOOTLOADER_BOSSA` Kconfig option as Y.  This will
+   automatically select :option:`CONFIG_USE_DT_CODE_PARTITION` Kconfig option.
+   Your board's :file:`.defconfig` file should define
+   :option:`CONFIG_BOOTLOADER_BOSSA_ARDUINO` or the
+   :option:`CONFIG_BOOTLOADER_BOSSA_ADAFRUIT_UF2` Kconfig option as Y.  This
+   selects the proper compatible SAM-BA bootloader mode.
+3. You need build and flash the SAM-BA bootloader for your device
```
Similar comments here as for the previous section

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+Typical flash layout and configuration
+--------------------------------------
+
+For those bootloaders that resides on flash, the device tree partitions layout
```
```suggestion
For bootloaders that reside on flash, the devicetree partition layout
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+--------------------------------------
+
+For those bootloaders that resides on flash, the device tree partitions layout
+is mandatory.  For the devices that have ROM bootloader this should be
```
```suggestion
is mandatory.  For devices that have a ROM bootloader, they are mandatory when
the application uses a storage or other non-application partition.
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+For those bootloaders that resides on flash, the device tree partitions layout
+is mandatory.  For the devices that have ROM bootloader this should be
+necessary when application demands the storage_partition.  In this special
+case, the boot partition will be omitted and code_partition will starts from
```
```suggestion
case, the boot partition should be omitted and code_partition should start from
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+is mandatory.  For the devices that have ROM bootloader this should be
+necessary when application demands the storage_partition.  In this special
+case, the boot partition will be omitted and code_partition will starts from
+offset 0.  It is necessary define the partitions with right size to avoid
```
```suggestion
offset 0.  It is necessary to define the partitions with sizes that avoid
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+offset 0.  It is necessary define the partitions with right size to avoid
+overlaps, always.
+
+Typical flash layout for devices without ROM bootloader
```
```suggestion
A typical flash layout for devices without a ROM bootloader is:
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+		};
+	};
+
+Typical flash layout for devices with ROM bootloader and storage
```
```suggestion
A typical flash layout for devices with a ROM bootloader and storage partition is:
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,151 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
+		};
+	};
+
+More implementation detailed can be found at boards documentation.  As quick
+reference, these four boards configurations can be explored.
```
```suggestion
More implementation details can be found in the :ref:`boards` documentation.  As a quick
reference, see these four board documentation pages:
```

mbolivar-nordic reviewed (changes requested):
Looks good, just a bunch of minor things.

nandojve commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -96,27 +83,65 @@ def is_extended_samba_protocol(self):
+    def get_board_name(self):
+        build_conf = BuildConfiguration(self.cfg.build_dir)
+
+        if 'CONFIG_BOARD' not in build_conf:
+            return '<board>'
```
It is not a error handler. It is used to print the board name at an error handler to user and I simple added a fallback.

mbolivar-nordic replied:
> It is not a error handler. It is used to print the board name at an error handler to user and I simple added a fallback.

I understand the code. What I'm saying is that if CONFIG_BOARD is missing, the build configuration is **really** messed up. That's a much bigger problem than CONFIG_USE_DT_CODE_PARTITION not being set.

The error message is therefore misleading, because build without CONFIG_BOARD set at all is going to fail for many other bigger reasons. So I'm wondering if the error message could be improved.

But like I said, not a blocker.

nandojve commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -37,35 +42,17 @@ def capabilities(cls):
     def do_add_parser(cls, parser):
         parser.add_argument('--bossac', default='bossac',
                             help='path to bossac, default is bossac')
-        parser.add_argument('--offset', default=None,
-                            help='start erase/write/read/verify operation '
-                                 'at flash OFFSET; OFFSET must be aligned '
-                                 ' to a flash page boundary')
```
I prefer have options at command line but I don't see why keep this. It not help and can create confusion.

1. In a baremetal environment, for instance, the offset must have a entry on the linker, otherwise it will not work. If we keep this option, the "linker" must be defined anyway on devicetree. As far I could understand it seams to be redundant.
2. In a multi image environment (secure/non-secure) there should be at least 2 offsets. west can get the offset from `.config` file and flash one images at time.

Do you see a scenario that may require --offset when invoke west at cmd line?

mbolivar-nordic replied:
> Do you see a scenario that may require --offset when invoke west at cmd line?

So you're saying the user should **never** provide this option, and it was always wrong to have it? If so, OK.

nandojve replied:
Yes, exactly!

nandojve commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -1,29 +1,34 @@
+from zephyr_ext_common import ZEPHYR_SCRIPTS
+
+# This is needed to load edt.pickle files.
+sys.path.append(str(ZEPHYR_SCRIPTS / 'dts'))
```
After add `import edtlib` the Compliance / Run compliance checks on patch series (PR) (pull_request) :
```
Run if [ -s Nits.txt ]; then
Error: See https://www.pylint.org/ for more details

************* Module runners.bossac
scripts/west_commands/runners/bossac.py:19:0: W0611: Unused import edtlib (unused-import)
Error: Process completed with exit code 1.
```

mbolivar-nordic replied:
> After add `import edtlib` the Compliance / Run compliance checks on patch series (PR) (pull_request) :

You can ignore this the same way sanitycheck does:

https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/sanity_chk/sanitylib.py#L66

```py
import edtlib  # pylint: disable=unused-import
```

mbolivar-nordic commented on doc/guides/device_mgmt/dfu.rst:
```diff
@@ -50,3 +50,169 @@ in the `MCUboot with Zephyr`_ documentation page on the MCUboot website.
 .. _MCUboot GitHub Project: https://github.com/runtimeco/mcuboot
+
+
+.. _atmel_sam_ba_bootloader:
```
Hey, I'm sorry about this but I didn't notice the first time I reviewed this PR that the documentation was going into this file.

I don't think it's really appropriate to put this here. This is about a west flash runner, not device management.

Can I suggest putting this in guides/debugging/host-tools.rst instead and generalizing that page to "flashing and debugging", since bossac only supports flashing?

@MaureenHelm -- would that work for you?

MaureenHelm replied:
Yes

nandojve replied:
Done!

mbolivar-nordic commented on doc/guides/debugging/index.rst:
```diff
@@ -1,7 +1,7 @@
-.. _debugging:
+.. _flashing_and_debugging:
```
Docs changes look good. Just one last thing related to the documentation: I think it would be better to move this directory to `guides/flash-debug` or so.

nandojve replied:
Sure, no problem!

mbolivar-nordic commented:
Built docs locally, they seem good to me. Thanks for bearing with the last minute change.

nandojve commented:
> Built docs locally, they seem good to me. Thanks for bearing with the last minute change.

No problem! Thank you for your review.

galak commented:
@MaureenHelm can you look at this for the doc changes.

galak commented:
@nandojve what happens with this code if you use it with SDK 0.11.4?

nandojve commented:
> @nandojve what happens with this code if you use it with SDK 0.11.4?

The code is supposed to be compatible with old SDK versions (< 0.12.0). I've been testing with 0.11.3 and it seems work as expected. The limitation is what the bossac can handle, I mean, SoC supported and protocols.

MaureenHelm commented:
> @MaureenHelm can you look at this for the doc changes.

Doc changes look fine, but I think we may want to add a redirect link in `doc/conf.py`

nandojve commented:
> > @MaureenHelm can you look at this for the doc changes.
> 
> Doc changes look fine, but I think we may want to add a redirect link in `doc/conf.py`

I did, it seems there is no way to move a directory. It should be file by file, right?

MaureenHelm commented:
> I did, it seems there is no way to move a directory. It should be file by file, right?

not sure. @carlescufi ?

mbolivar-nordic commented:
Maureen Helm <notifications@github.com> writes:

>> I did, it seems there is no way to move a directory. It should be file by file, right?
>
> not sure. @carlescufi ?

Looks like @nashif was the one to add the redirects extension.

Judging by the source, it looks like it's for pages only, not
directories.

nandojve committed:
doc: guides: flash_debug: host-tools: Add SAM-BA docs

Add SAM-BA bootloader documentation at Zephyr host-tools in
flash_debug guides.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
> Maureen Helm <notifications@github.com> writes:
> > I did, it seems there is no way to move a directory. It should be file by file, right? not sure. @carlescufi ?
> Looks like @nashif was the one to add the redirects extension. Judging by the source, it looks like it's for pages only, not directories.

I was wondering if this can move forward since there is no other solution.

nandojve commented:
Hello folks!
Is there anything missing here?

galak commented:
LGTM.  Will wait on @MaureenHelm comments on the doc redirection issue.
